### PR TITLE
Numba for loops interpolation

### PR DESCRIPTION
**Pull request by mlubej**

### CODE REVIEW

**Comment by mlubej** [comment]
Need to rename the linear interpolation tasks before merging to `develop`

**Comment by andrejburja** [comment]
There is also one test failing.

**Comment by mlubej** [comment]
> There is also one test failing.

Yeah, I noticed, it seems to be in coregistration, but locally all my tests pass...

**Comment by mlubej** [comment]
@andrejburja the renaming introduced failing unit tests for interpolation. I expected them to fail due to the different interpolation method, but the error actually seems to occur due to some code crashes.

File: features/eolearn/features/interpolation.py [hunk 1]
```diff
@@ -5,11 +5,14 @@
...
 import datetime as dt
 import inspect
+import calendar
 from functools import partial
 
 import dateutil
 import scipy.interpolate
 import numpy as np
+import numba
+from numba import prange
```
[the remark this answers is not in the document]
Comment (andrejburja) [reply features/eolearn/features/interpolation.py]: prange changed to numba.prange

File: features/eolearn/features/interpolation.py [hunk 2]
```diff
@@ -389,17 +392,184 @@ def execute(self, eopatch):
...
             super().__init__(feature, np.interp, **kwargs)
         else:
             super().__init__(feature, scipy.interpolate.interp1d, kind='linear', **kwargs)
 
 
+class LinearInterpolation(InterpolationTask):
+    """
+    Implements `eolearn.features.InterpolationTask` by using `numpy.interp` and @numb.jit(nopython=True)
+    """
+    def __init__(self, feature, parallel=False, **kwargs):
```
[the remark this answers is not in the document]
Comment (andrejburja) [reply features/eolearn/features/interpolation.py]: done

File: features/eolearn/features/interpolation.py [hunk 3]
```diff
@@ -389,17 +392,184 @@ def execute(self, eopatch):
...
+                last = np.where(max_time < resampled_times)[0]
+                if last.size:
+                    new_data[last[0]:] = np.nan
+
+            new_bands[:, n_feat] = new_data
+
+        return new_bands.reshape(len(resampled_times), height, width, depth)
+
+    @staticmethod
+    def datetime_to_utc(datetimes, start=None):
```
[the remark this answers is not in the document]
Comment (andrejburja) [reply features/eolearn/features/interpolation.py]: Method deleted.

File: features/eolearn/features/interpolation.py [hunk 4]
```diff
@@ -389,17 +392,184 @@ def execute(self, eopatch):
...
+                if last.size:
+                    new_data[last[0]:] = np.nan
+
+            new_bands[:, n_feat] = new_data
+
+        return new_bands.reshape(len(resampled_times), height, width, depth).astype(data.dtype)
+
+    @staticmethod
+    @numba.jit(nopython=True, parallel=True)
+    def interpolation_function_parallel(data, times, resampled_times):
```
[the remark this answers is not in the document]
Comment (andrejburja) [reply features/eolearn/features/interpolation.py]: I created `base_interpolation_function` and added 2 functions decorated with `numba.njit` - one regular, one parallel
all 3 functions are at the beginning of file, outside class

File: features/eolearn/features/interpolation.py [hunk 5]
```diff
@@ -389,17 +392,184 @@ def execute(self, eopatch):
...
+        :return: Array of interpolated values
+        :rtype: numpy.ndarray
+        """
+
+        timestamps, height, width, depth = data.shape
+        new_bands = np.empty((len(resampled_times), height * width * depth))
+        data = data.reshape(timestamps, height * width * depth)
+        for n_feat in range(height * width * depth):
+            mask1d = ~np.isnan(data[:, n_feat])
+            if (~mask1d).all():
```
[the remark this answers is not in the document]
Comment (andrejburja) [reply features/eolearn/features/interpolation.py]: ok

File: features/requirements.txt [hunk 6]
```diff
@@ -3,3 +3,4 @@ numpy
 scipy
 python-dateutil
 scikit-image
+numba>=0.43
```
[the remark this answers is not in the document]
Comment (andrejburja) [reply features/requirements.txt]: I hope it is not going to change in the future.

File: features/eolearn/features/interpolation.py [hunk 7]
```diff
@@ -389,17 +392,184 @@ def execute(self, eopatch):
...
+                new_data = np.empty((len(resampled_times)))
```
[the remark this answers is not in the document]
Comment (andrejburja) [reply features/eolearn/features/interpolation.py]: noted

File: features/eolearn/tests/test_interpolation.py [hunk 8]
```diff
@@ -1,12 +1,11 @@
...
 from datetime import datetime
+import numpy as np
 
 from eolearn.core import EOPatch, FeatureType
 from eolearn.features import LinearInterpolation, CubicInterpolation, SplineInterpolation, BSplineInterpolation, \
-    AkimaInterpolation, LinearResampling, CubicResampling, NearestResampling, KrigingInterpolation
+    AkimaInterpolation, LinearResampling, CubicResampling, NearestResampling, KrigingInterpolation, LegacyInterpolation
 
 
 class TestInterpolation(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (andrejburja) [reply features/eolearn/tests/test_interpolation.py]: Added `parallel=True` to one test.

File: features/eolearn/features/interpolation.py [hunk 9]
```diff
@@ -389,17 +392,184 @@ def execute(self, eopatch):
...
 
+class LinearInterpolation(InterpolationTask):
+    """
+    Implements `eolearn.features.InterpolationTask` by using `numpy.interp` and @numb.jit(nopython=True)
+    """
+    def __init__(self, feature, parallel=False, **kwargs):
+        super().__init__(feature, np.interp, **kwargs)
+        self.parallel = parallel
+
+    def execute(self, eopatch):
```
[the remark this answers is not in the document]
Comment (andrejburja) [reply features/eolearn/features/interpolation.py]: execute() function is the same for both cases, i just added interpolate_data() to LinearInterpolation

File: features/eolearn/features/interpolation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AleksMat) [reply features/eolearn/features/interpolation.py]: I see you moved `parallel` parameter to `InterpolationTask`. However this parameter is only specific to `LinearInterpolation` and therefore should be defined at `LinearInterpolation` class as before.

**Pull request merged**